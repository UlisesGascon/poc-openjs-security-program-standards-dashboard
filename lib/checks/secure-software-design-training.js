import { getConfig } from '../../config/index.js'
import { isDateWithinPolicy } from '../utils/index.js'

const { generalExpirationPolicy } = getConfig()

export function checkSecureSoftwareDesignTraining (projects, checks) {
  checks.secureSoftwareDesignTraining = {
    name: 'Secure Software Design Training',
    description: 'At least One Primary Maintainer has taken TBD Training on Secure Software Design',
    level: {
      incubating: 'expected',
      graduated: 'expected',
      retiring: 'expected'
    },
    url: 'https://openjs-security-program-standards.netlify.app/details/item-0',
    projects: {},
    category: 'code quality'
  }

  for (const project of projects) {
    checks.secureSoftwareDesignTraining.projects[project.name] = {}
    const trainingDetails = project.additionalInformation?.secureSoftwareDesignTraining || {}

    const criteria = trainingDetails.globalAchievement &&
            isDateWithinPolicy(trainingDetails.confirmationDate, generalExpirationPolicy)

    if (criteria) {
      checks.secureSoftwareDesignTraining.projects[project.name].globalAchievement = true
      checks.secureSoftwareDesignTraining.projects[project.name].rationale = trainingDetails.rationale
    } else {
      checks.secureSoftwareDesignTraining.projects[project.name].globalAchievement = false
      checks.secureSoftwareDesignTraining.projects[project.name].rationale = 'The team has not completed the required training or requires an update.'
    }
  }
}
