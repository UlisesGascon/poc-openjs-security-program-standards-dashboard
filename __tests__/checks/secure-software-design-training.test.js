import { checkSecureSoftwareDesignTraining } from '../../lib/checks/secure-software-design-training.js'
import { isDateWithinPolicy } from '../../lib/utils/index.js'

jest.mock('../../lib/utils/index.js')

describe('checkSecureSoftwareDesignTraining', () => {
  let projects
  let checks
  let tasks
  let alerts

  beforeEach(() => {
    projects = [
      {
        name: 'Project1',
        additionalInformation: {
          secureSoftwareDesignTraining: {
            globalAchievement: true,
            rationale: 'Training completed',
            confirmationDate: new Date().toISOString()
          }
        }
      },
      {
        name: 'Project2',
        additionalInformation: {
          secureSoftwareDesignTraining: {
            globalAchievement: false,
            rationale: 'Training not completed',
            confirmationDate: new Date().toISOString()
          }
        }
      }
    ]
    checks = {}
    tasks = {}
    alerts = {}
    isDateWithinPolicy.mockReturnValue(true)
  })

  it('should update checks with project compliance data', () => {
    checkSecureSoftwareDesignTraining({ projects, checks, tasks, alerts })
    expect(checks).toMatchSnapshot()
    expect(tasks).toMatchSnapshot()
    expect(alerts).toMatchSnapshot()
  })

  it('should handle projects with no additional information', () => {
    projects = [
      {
        name: 'Project3'
      }
    ]
    checkSecureSoftwareDesignTraining({ projects, checks, tasks, alerts })
    expect(checks).toMatchSnapshot()
    expect(tasks).toMatchSnapshot()
    expect(alerts).toMatchSnapshot()
  })

  it('should handle projects with expired training', () => {
    projects = [
      {
        name: 'Project4',
        additionalInformation: {
          secureSoftwareDesignTraining: {
            globalAchievement: true,
            rationale: 'Training completed',
            confirmationDate: new Date().toISOString()
          }
        }
      }
    ]
    isDateWithinPolicy.mockReturnValue(false)
    checkSecureSoftwareDesignTraining({ projects, checks, tasks, alerts })
    expect(checks).toMatchSnapshot()
    expect(tasks).toMatchSnapshot()
    expect(alerts).toMatchSnapshot()
  })
})
