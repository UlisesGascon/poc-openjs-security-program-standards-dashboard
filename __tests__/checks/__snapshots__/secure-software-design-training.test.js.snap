// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`checkSecureSoftwareDesignTraining should handle projects with expired training 1`] = `
{
  "secureSoftwareDesignTraining": {
    "category": "code quality",
    "description": "At least One Primary Maintainer has taken TBD Training on Secure Software Design",
    "level": {
      "graduated": "expected",
      "incubating": "expected",
      "retiring": "expected",
    },
    "name": "Secure Software Design Training",
    "projects": {
      "Project4": {
        "globalAchievement": false,
        "rationale": "The team has not completed the required training or requires an update.",
      },
    },
    "url": "https://openjs-security-program-standards.netlify.app/details/item-0",
  },
}
`;

exports[`checkSecureSoftwareDesignTraining should handle projects with no additional information 1`] = `
{
  "secureSoftwareDesignTraining": {
    "category": "code quality",
    "description": "At least One Primary Maintainer has taken TBD Training on Secure Software Design",
    "level": {
      "graduated": "expected",
      "incubating": "expected",
      "retiring": "expected",
    },
    "name": "Secure Software Design Training",
    "projects": {
      "Project3": {
        "globalAchievement": false,
        "rationale": "The team has not completed the required training or requires an update.",
      },
    },
    "url": "https://openjs-security-program-standards.netlify.app/details/item-0",
  },
}
`;

exports[`checkSecureSoftwareDesignTraining should update checks with project compliance data 1`] = `
{
  "secureSoftwareDesignTraining": {
    "category": "code quality",
    "description": "At least One Primary Maintainer has taken TBD Training on Secure Software Design",
    "level": {
      "graduated": "expected",
      "incubating": "expected",
      "retiring": "expected",
    },
    "name": "Secure Software Design Training",
    "projects": {
      "Project1": {
        "globalAchievement": true,
        "rationale": "Training completed",
      },
      "Project2": {
        "globalAchievement": false,
        "rationale": "The team has not completed the required training or requires an update.",
      },
    },
    "url": "https://openjs-security-program-standards.netlify.app/details/item-0",
  },
}
`;
